### PR TITLE
Qat pytorch finalize

### PR DESCRIPTION
**Pull request by elad-c**

### CODE REVIEW

File: model_compression_toolkit/qat/pytorch/quantizer/ste_rounding/uniform_ste.py [hunk 1]
```diff
@@ -96,7 +98,23 @@ def __call__(self,
...
+        return uniform_quantizer(inputs, self.quantizer_parameters[FQ_MIN], self.quantizer_parameters[FQ_MAX], self.num_bits)
+
+    def convert2inferable(self) -> iq.WeightsUniformInferableQuantizer:
+        """
+        Convert quantizer to inferable quantizer.
+
+        Returns:
+            A pytorch inferable quanizer object.
+        """
+        _min = self.quantizer_parameters[FQ_MIN] #.cpu().detach().numpy()
```
Comment (eladc-git) [inline model_compression_toolkit/qat/pytorch/quantizer/ste_rounding/uniform_ste.py:110]: what about .cpu().detach().numpy() (the comment lines)?

File: model_compression_toolkit/qat/pytorch/quantizer/ste_rounding/uniform_ste.py [hunk 2]
```diff
@@ -96,7 +98,23 @@ def __call__(self,
...
+        _max = self.quantizer_parameters[FQ_MAX] #.cpu().detach().numpy()
+
+        print(f' ==> Inferable Quantizer: {_min}, {_max}, {self.num_bits}')
```
Comment (eladc-git) [inline model_compression_toolkit/qat/pytorch/quantizer/ste_rounding/uniform_ste.py:113]: print?

File: model_compression_toolkit/quantizers_infrastructure/keras/quantize_wrapper.py [hunk 3]
```diff
@@ -120,7 +120,7 @@ def build(self, input_shape):
                 trainable=False)
 
             self._weight_vars = []
-            for name, quantizer in self.dispatcher.weight_quantizers.items():
+            for name, quantizer in self._dispatcher.weight_quantizers.items():
```
Comment (eladc-git) [inline model_compression_toolkit/quantizers_infrastructure/keras/quantize_wrapper.py:123]: why changing keras in pytorch PR?

**Review by eladc-git — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py [hunk 4]
```diff
@@ -41,17 +44,53 @@ def __init__(self,
...
+
+        def set_weight_quantizers(self, weight_quantizers: List[BaseInferableQuantizer]):
+            self._dispatcher.set_weight_quantizers(weight_quantizers)
+            self._set_weight_quantizers()
+
+        def set_activation_quantizers(self, activation_quantizers: List[BaseInferableQuantizer]):
+            self._dispatcher.set_activation_quantizers(activation_quantizers)
+            self._set_activation_quantizers()
+
+        def _set_weight_quantizers(self):
```
Comment (eladc-git) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py:89]: function deccription

File: model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py [hunk 5]
```diff
@@ -41,17 +44,53 @@ def __init__(self,
...
+            if self._dispatcher.is_weights_quantization:
+                inferable_weight_quantizers = {}
+                for name, quantizer in self._dispatcher.weight_quantizers.items():
+                    if isinstance(quantizer, qi.BasePytorchTrainableQuantizer):
+                        inferable_weight_quantizers.update({name: quantizer.convert2inferable()})
+                    else:
+                        Logger.error('Can only convert trainable quantizers based on BasePytorchTrainableQuantizer')
+                self.set_weight_quantizers(inferable_weight_quantizers)
+
+        def set_weight_quantizers(self, weight_quantizers: List[BaseInferableQuantizer]):
```
Comment (eladc-git) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py:81]: function deccription

File: model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py [hunk 6]
```diff
@@ -41,17 +44,53 @@ def __init__(self,
...
                 # Functional layers
                 setattr(self, LAYER, module)
 
+            # Init weight quantizers from dispatcher
+            self._set_weight_quantizers()
+
+            # Init activations quantizers from dispatcher
+            self._set_activation_quantizers()
+
+        def convert_to_inferable_quantizers(self):
```
Comment (eladc-git) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py:60]: function description

File: model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py [hunk 7]
```diff
@@ -41,17 +44,53 @@ def __init__(self,
...
+        def set_activation_quantizers(self, activation_quantizers: List[BaseInferableQuantizer]):
+            self._dispatcher.set_activation_quantizers(activation_quantizers)
+            self._set_activation_quantizers()
+
+        def _set_weight_quantizers(self):
             self._weight_vars = []
 
             # Init weights quantizers
-            for name, quantizer in self.dispatcher.weight_quantizers.items():
+            for name, quantizer in self._dispatcher.weight_quantizers.items():
```
Comment (eladc-git) [inline model_compression_toolkit/quantizers_infrastructure/pytorch/quantize_wrapper.py:93]: This mean you run "initialize_quantization" and  "weight.detach()" to inferable quantizers

**Review by eladc-git — CHANGES_REQUESTED** [review]

**Review by eladc-git — APPROVED** [review]
